### PR TITLE
Prune EvalModelTemplate from tests

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Here updated all the tests that were using the old `EvalModelTemplate`. Will prune it and its components in a follow-up.
Part of #4501

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -99,13 +125,8 @@ def test_dataloader_config_errors_init(tmpdir, dataloader_options):
 def test_multiple_val_dataloader(tmpdir):
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:125]: I'd suggest removing this test entirely... It's not testing anything useful really

File: tests/trainer/test_dataloaders.py [hunk 2]
```diff
@@ -114,58 +135,34 @@ def test_multiple_val_dataloader(tmpdir):
     # verify there are 2 val loaders
     assert len(trainer.val_dataloaders) == 2, "Multiple val_dataloaders not initiated properly"
 
-    # make sure predictions are good for each val set
-    for dataloader in trainer.val_dataloaders:
-        tpipes.run_prediction_eval_model_template(trained_model=model, dataloader=dataloader)
-
 
 @pytest.mark.parametrize("ckpt_path", [None, "best", "specific"])
 def test_multiple_eval_dataloader(tmpdir, ckpt_path):
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:140]: Same for this test

File: tests/trainer/test_dataloaders.py [hunk 3]
```diff
@@ -114,58 +135,34 @@ def test_multiple_val_dataloader(tmpdir):
...
 
     # only train passed to fit
-    model = EvalModelTemplate()
-    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, limit_val_batches=0.1, limit_train_batches=0.2)
-    fit_options = dict(train_dataloaders=model.dataloader(train=True))
+    model = BoringModel()
+    train_loader = model.train_dataloader()
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=2)
+    fit_options = dict(train_dataloaders=train_loader)
     trainer.fit(model, **fit_options)
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:163]:
```suggestion
    trainer.fit(model, train_dataloaders=train_loader)
```

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -96,42 +101,8 @@ def test_no_val_module(monkeypatch, tmpdir, tmpdir_server, url_ckpt):
         if url_ckpt
         else new_weights_path
     )
-    model_2 = EvalModelTemplate.load_from_checkpoint(checkpoint_path=ckpt_path, hparams_file=hparams_path)
-    model_2.eval()
-
-
-@pytest.mark.parametrize("url_ckpt", [True, False])
-def test_no_val_end_module(monkeypatch, tmpdir, tmpdir_server, url_ckpt):
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:104]: Why is this test gone?

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -553,14 +537,21 @@ def test_trainer_max_steps_and_epochs_fit_loop_done(max_epochs, max_steps, is_do
...
     """Verify model trains according to specified min steps."""
-    model = EvalModelTemplate()
-    num_train_samples = math.floor(len(model.train_dataloader()) * 0.5)
+    num_train_samples = math.floor(len(BoringModel().train_dataloader()) * 0.5)
+
+    class CustomModel(BoringModel):
+        def training_step(self, *args, **kwargs):
+            # try to force stop right after first step
+            if self.global_step > 0:
+                self.trainer.should_step = True
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:546]:
```suggestion
                self.trainer.should_stop = True
```

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -1128,9 +1129,16 @@ def test_num_sanity_val_steps(tmpdir, limit_val_batches):
 def test_num_sanity_val_steps_neg_one(tmpdir, limit_val_batches):
     """Test that `num_sanity_val_steps=-1` runs through all validation data once, and as many batches as limited by
     `limit_val_batches` Trainer argument."""
-    model = EvalModelTemplate()
-    model.validation_step = model.validation_step__multiple_dataloaders
-    model.validation_epoch_end = model.validation_epoch_end__multiple_dataloaders
+
+    class CustomModel(BoringModel):
+        def validation_step(self, batch, batch_idx, dataloader_idx):
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1134]: I think we could add `dataloader_idx=0` to BoringModel so this isn't necessary in each subclass

**Review by carmocca — APPROVED** [review]
Great work!

I'll push a commit removing the dead code

**Comment by rohitgr7** [comment]
> I'll push a commit removing the dead code

I was thinking to do that in a separate one to make things easy for the reviewer.

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (rohitgr7) [reply tests/trainer/test_trainer.py:104]: not sure why this was there. Check the test above. Both are exactly the same.

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (rohitgr7) [reply tests/trainer/test_dataloaders.py:125]: yeah.. must be very old.

**Comment by carmocca** [comment]
> I was thinking to do that in a separate one to make things easy for the reviewer. There's a lot of dead code now.

Deleted code is easy to review. I only removed the functions that are no longer used, but no the classes entirely. If you prefer, you can revert the commit

File: tests/helpers/boring_model.py [hunk 7]
```diff
@@ -100,7 +100,7 @@ def step(self, x):
         out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
         return out
 
-    def training_step(self, batch, batch_idx):
+    def training_step(self, batch, batch_idx, dataloader_idx=0):
```
Comment (awaelchli) [inline tests/helpers/boring_model.py:103]: Pretty sure this should not be added here. 99% of time we won't have multiple dataloaders. There may be specific tests that can add this in an override.

File: tests/trainer/test_dataloaders.py [hunk 8]
```diff
@@ -96,76 +116,16 @@ def test_dataloader_config_errors_init(tmpdir, dataloader_options):
         Trainer(default_root_dir=tmpdir, max_epochs=1, **dataloader_options)
 
 
-def test_multiple_val_dataloader(tmpdir):
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:99]: it's extremely hard to review, because it's not clear where things are being moved.
Please consider only converting EvalModelTemplate and do the moving of tests in another PR.

File: tests/trainer/test_dataloaders.py [hunk 9]
```diff
@@ -458,13 +410,7 @@ def test_dataloaders_with_limit_percent_batches(tmpdir, limit_train_batches, lim
 def test_dataloaders_with_limit_num_batches(tmpdir, limit_train_batches, limit_val_batches, limit_test_batches):
     """Verify num_batches for train, val & test dataloaders passed with batch limit as number."""
 
-    model = EvalModelTemplate()
```
Comment (tchaton) [inline tests/trainer/test_dataloaders.py:461]: Neat! This was quite ugly ...

**Review by tchaton — COMMENTED** [review]
Do you see any way to split this PR further down, it is a bit hard to fully trace all changes and ensure we don't introduce a regression.

**Comment by rohitgr7** [comment]
@tchaton sure! doing it module by module.
